crates/iceberg/src/spec/manifest.rs

```diff
@@ -133,9 +133,9 @@ pub struct ManifestWriter {
 
 impl ManifestWriter {
     /// Create a new manifest writer.
-    pub fn new(output: OutputFile, snapshot_id: i64, key_metadata: Vec<u8>) -> Self {
+    pub fn new(manifest_path: String, snapshot_id: i64, key_metadata: Vec<u8>) -> Self {
```

Comment:
> In my mind the manifest path "metadata" and where the serialisation actually occurs (e.g. in S3) will always be the same. I.e. we write to s3://metadata/.../v1.table-metadata.json, which is both the place where it is serialised to a file and a metadata pointer of "where is this file"

So - yes. As part of the spec that _is_ a requirement; and `location` is both where the data gets serialized to and is included in the returned `ManifestFile`.

The problem here is that `iceberg-rust` enforces that by passing in the `OutputFile` to a new `ManifestWriter`; which the writer then puts in the `ManifestFile` returned by `writer.write(manifest)`  _this is the coupling of serialization and building._ Because of this, the user **cannot** build a metadata object (create the Rust type of that object) with `location = /tmp/foo.avro` and serialize it (write that Rust type as bytes) to `/tmp/bar.avro`. The `iceberg-rust` crate handles that for you. So if I pass in an `OutputFile` with `location` `/tmp/foo.avro` to a `ManifestWriter` _that's where the `ManifestFile` is being serialized to whether I like it or not_.

This is why we can't use `ObjectStore` at the moment. If we were to build/serialize the bytes on a node `A` and then write those bytes to node `B`, _and query them from node `B`_; the queries will fail because the metadata that has been written to node `B` will point to metadata on node `A`. Because that's where the building/serialization took place. And those take place together because the `iceberg-rust` crate has coupled them together.

IMO, Building/serialization needs to be be made completely separate in the `iceberg-rust` crate. As an example, for `ManifestFile` you would ideally have some method, say `ManifestFile::to_bytes(self) -> Bytes` that consumes the `ManifestFile` and serializes it. Where those bytes go is entirely up to the user, _but the user is required to uphold the invariant that where those bytes get written to match the [`manifest_path`](https://docs.rs/iceberg/latest/iceberg/spec/struct.ManifestFile.html#structfield.manifest_path) for the `ManifestFile` that got serialized. However the maintainers [seem keen on preserving that coupling](https://github.com/apache/iceberg-rust/issues/778#issuecomment-2538270065); but want to change the `FileIO` type to be a trait so that users can provide their own object_store/storage implementations.

Part of me wonders if we should try to convince them to separate things out better; but I'm not sure how worthy that is given that a `FileIO` trait would allow us to implement that trait for `Arc<dyn ObjectStore>`, which _should_ meet the design goal of using `ObjectStore`.

Reply:
> The problem here is that iceberg-rust enforces that by passing in the OutputFile to a new ManifestWriter

Note: This PR **_removes_** that enforcement - hence why I think having a discussion about it first on this PR is important.